File: 1457-pseudo-palindromic-paths-in-a-binary-tree/1457-pseudo-palindromic-paths-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pseudoPalindromicPaths (self, root: Optional[TreeNode]) -> int:
        ans = 0
        def dfs(root, st):
            if (not root.left) and (not root.right):
                nonlocal ans
                if root.val in st: st[root.val] += 1
                else: st[root.val]=1
                t=0
                for i in st:
                    if st[i]%2!=0: t+=1
                if t<2: ans+=1
                return
            if root.val in st: st[root.val] += 1
            else: st[root.val] = 1
            st2=st.copy()
            if root.left: dfs(root.left, st)
            if root.right: dfs(root.right, st2)
            return
        dfs(root, {})
        return ans

File: 1457-pseudo-palindromic-paths-in-a-binary-tree/1457-pseudo-palindromic-paths-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */


class Solution {
    public int pseudoPalindromicPaths (TreeNode root) {
        return dfs(root, new boolean[10]);
    }
    int dfs(TreeNode root, boolean[] map){
        if(root.left==null && root.right==null){
            map[root.val] ^= true;
            int odd = 0;
            for(var m: map) odd += m ? 1 : 0;
            map[root.val] ^= true;
            return odd < 2 ? 1 : 0;
        }
        map[root.val] ^= true;
        int ans = 0;
        if(root.left != null) ans += dfs(root.left, map);
        if(root.right != null) ans += dfs(root.right, map);
        map[root.val] ^= true;
        return ans;
    }
}





// class Solution {
//     public int pseudoPalindromicPaths (TreeNode root) {
//         return dfs(root, new HashMap<>());
//     }
//     int dfs(TreeNode root, Map<Integer, Integer> map){
//         if(root.left==null && root.right==null){
//             map.put(root.val, map.getOrDefault(root.val, 0)+1);
//             // System.out.println(map);
//             int odd = 0;
//             for(int key: map.keySet()){
//                 odd += map.get(key)%2;
//             }
//             map.put(root.val, map.get(root.val)-1);
//             return odd < 2 ? 1 : 0;
//         }
//         map.put(root.val, map.getOrDefault(root.val, 0)+1);
//         int ans = 0;
//         if(root.left != null) ans += dfs(root.left, map);
//         if(root.right != null) ans += dfs(root.right, map);
//         map.put(root.val, map.get(root.val)-1);
//         return ans;
//     }
// }